[V1][TPU] TPU-optimized top-p implementation (avoids scattering).

hyeygit opened the pull request:
Top-k and top-p are slow on TPU because existing algorithms use `torch.scatter`. For some reason `torch.scatter` is extremely slow on TPU. There's ongoing work to optimize it, but until that's done, we need an alternative algorithm that circumvents scattering.

The algorithm in this PR avoids torch.scatter by finding a "cut-off" element in the original logit, and after thresholding the logit using this cut-off, the remaining elements shall constitute the top-p set. This is inspired by the `apply_top_k_only` algorithm created by @njhill in #15478.

### Benchmark

Microbenchmark (on v6e-1) shows significant speed up -- "Running 32 elapsed time" is ~5 ms, down from the original scatter-based algorithm's ~500 ms, a **100x improvement**.

<details>

<summary>Microbenchmark full results on v6e-1</summary>

```
$ VLLM_USE_V1=1 python sampler_microbenchmark.py 
INFO 03-31 14:40:10 [__init__.py:239] Automatically detected platform tpu.
WARNING:root:libtpu.so and TPU device found. Setting PJRT_DEVICE=TPU.
INFO 03-31 14:40:15 [topk_topp_sampler.py:82] Using approximate top-p optimized for TPU. Result may in theory differ from the exact algorithm if there are tokens with near-identical probabilities (< 1e-9 diff).
Compiling/Warmup 1 elapsed time: 9.270433902740479
Compiling/Warmup 4 elapsed time: 9.104885816574097
Compiling/Warmup 16 elapsed time: 8.811976194381714
Compiling/Warmup 32 elapsed time: 8.926635026931763
Running 1 elapsed time: 0.004515171051025391
Running 1 elapsed time: 0.003937482833862305
Running 1 elapsed time: 0.003930091857910156
Running 1 elapsed time: 0.0038993358612060547
Average time:  0.0040705204010009766
Running 4 elapsed time: 0.0042819976806640625
Running 4 elapsed time: 0.004051685333251953
Running 4 elapsed time: 0.00403141975402832
Running 4 elapsed time: 0.004080057144165039
Average time:  0.004111289978027344
Running 16 elapsed time: 0.00475311279296875
Running 16 elapsed time: 0.0045032501220703125
Running 16 elapsed time: 0.0044858455657958984
Running 16 elapsed time: 0.19616365432739258
Average time:  0.052476465702056885
Running 32 elapsed time: 0.00586247444152832
Running 32 elapsed time: 0.005380868911743164
Running 32 elapsed time: 0.005262851715087891
Running 32 elapsed time: 0.0052433013916015625
Average time:  0.005437374114990234
```

</details>

### Extra notes

The `VLLM_TPU_DISABLE_TOPK_TOPP_OPTIMIZATION` env (introduced in #15242) can now be removed. Not done in this PR since @NickLucche's pending PR #15489 already handles it (thanks!).

Commit:
[V1][TPU] Add a TPU-optimized top-p algorithm.

Top-k and top-p are slow on TPU because existing algorithms use torch.scatter.
For some reason torch.scatter is extremely slow on TPU. There's ongoing work
to optimize it, but until that's done, we need an alternative algorithm that
circumvents scattering.

The algorithm in this PR avoids using torch.scatter by finding a "cut-off"
element in the original logit, and after thresholding the logit using this
cut-off, the remaining elements shall constitute the top-p set.

A caveat of the above approach is that ties are not correctly handled --
if there are duplicate cutoff elements present in the logit, then the
resulting top-p set will be incorrect. To address this problem, we
introduce a tiny perturbation to the probabilities (after softmax) to
break any potential ties. The added perturbation is tiny so it should
not alter the end results significantly, but it still makes this algorithm
approximate rather than an exact one.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

brittrock commented on vllm/v1/sample/ops/topk_topp_sampler.py:
```diff
@@ -122,23 +122,65 @@ def forward_tpu(
-            # logits = apply_top_k_top_p(logits, k, p)
-            pass
-
+        logits = apply_top_k_top_p_tpu(logits, k, p)
```
Thanks @hyeygit , this seems reasonable to me in the interim but I'll let the other folks chime in on the appropriateness. Cc @yaochengji @yarongmu-google 

Given this can slightly impact the generated output during ties, this really feels like something we should be warning the user about. Not every time the function is called, of course, but at a minimum, we should be warning users when the  argument is set to anything other than the default. I couldn't find a warning log but I'm also on my phone, so apologies if I just missed it.

hyeygit replied:
Thanks for the review @brittrock . I added a one-time log message about this algorithm being approx in theory.

In practice, I think the tiny 1e-9 probability perturbation doesn't alter the result in any meaningful way. The _only_ situation where the output differs from the exact algo is if there are multiple tokens whose probabilities are within 1e-9 (one in a billion) of each other. This means they practically have the same probability, so including either one of them in the top-p set should be acceptable.

bvrockwell replied:
thanks for adding! 

I agree in practice, probably ok, but this could break accuracy tests and so good idea to include in any case. 

nice job, again!

bvrockwell replied:
I signed in from my phone and must have created another github account >.<

ignore my alter ego's request for review @hyeygit 😆

brittrock reviewed (changes requested):
Great work, thank you for adding this @hyeygit, including the notes on expected speedups. I just have minor usability feedback :)

NickLucche commented on tests/v1/tpu/test_topk_topp_sampler.py:
```diff
@@ -0,0 +1,120 @@
+if current_platform.is_tpu():
+    import torch_xla.core.xla_model as xm
+
+DEVICE = xm.xla_device() if current_platform.is_tpu() else torch.device("cuda")
```
I think if the test is under `v1/tpu` we shouldn't test for CUDA but skip if platform is not tpu. Otherwise we can move the test into the shared directory.

hyeygit replied:
Yep makes sense. Updated to TPU only.

NickLucche commented on tests/v1/tpu/test_topk_topp_sampler.py:
```diff
@@ -0,0 +1,120 @@
+
+        # Top-p + top-k
+        no_op_top_p = torch.tensor([1.0])
+        result2 = apply_top_k_top_p_tpu(logits=logits.clone(),
```
nice test!

hyeygit replied:
Credit to @njhill's apply_top_k_only test!

NickLucche commented on tests/v1/tpu/test_topk_topp_sampler.py:
```diff
@@ -0,0 +1,120 @@
+# SPDX-License-Identifier: Apache-2.0
```
can we add this test to `run-tpu-v1-test.sh`?

hyeygit replied:
Done.

NickLucche commented on vllm/v1/sample/ops/topk_topp_sampler.py:
```diff
@@ -122,23 +126,65 @@ def forward_tpu(
+    if k is not None:
+        logits = apply_top_k_only(logits, k)
+
+    if p is not None:
+        logits = apply_approx_top_p(logits, p)
```
is this top-p (to be used as topp-only) implementation also needed on gpu? @njhill

hyeygit replied:
IIUC scattering isn't a bottleneck on GPU so this impl wouldn't bring much benefit (plus, this impl still involves a full vocab sort same as the forward_native version).

NickLucche reviewed (changes requested):
Great job here @hyeygit ! Left some comments about tests.

Please remember to enable topp like I've done for k here https://github.com/vllm-project/vllm/pull/15489/files. 
Otherwise I can enable both in the same PR.

hyeygit committed:
Bugfix: Make sure uniform sampling happens on XLA not CPU.

Previously the uniform sampling happens on CPU and resulted in a slowdown.
(Running 32 elasped time is 39 ms). Moving the sampling to XLA sped things
up significantly, the benchmark time is down to 5 ms.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

hyeygit committed:
Update tests.

* Added a more comprehensive correctness test for top-p.
* Included tests/v1/tpu/test_topk_topp_sampler.py in run-tpu-v1-test.sh.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

hyeygit commented:
> Please remember to enable topp like I've done for k here https://github.com/vllm-project/vllm/pull/15489/files. Otherwise I can enable both in the same PR.

It might be cleaner to enable top-p in your PR (perhaps rename the PR to Enable both Top-k and Top-p) since it's blocked by this and the to-be-sent-out top-k PR. Let me know if that sounds alright! @NickLucche

NickLucche commented:
@hyeygit  Works for me! 
I may still do the enablement one at a time to track benchmarks.

njhill commented:
Thanks @hyeygit this looks good. However I don't really understand the need for the random perturbation / tiebreaking.

The behaviour without doing this will already be effectively the same as an arbitrary tie-break. The random pertubation doesn't even make it deterministic. If we want it to be deterministic we can set `stable=True` in the sort operation (but not suggesting that's needed).

Also this implicit tie-breaking behaviour is the same in the existing implementation, I don't see how it's something specific to the new one.

Re your related [comment](https://github.com/vllm-project/vllm/pull/15478#issuecomment-2764706287) on the top_k PR:
> However I think one corner case where this would break is if there are duplicate elements in the logit that equal the cut off value (i.e. `top_k_mask`). For example, given an input of `[1, 2, 2, 2, 3]` and `k=3`, the current `apply_top_k_only` would return `[-inf, 2, 2, 2, 3]` while the correct result should be `[-inf, -inf, 2, 2, 3]`.

Again this is no different to how the pre-existing impl works. And I think it's ok for the actual shortlist to comprise more than k tokens in the case that there's a tie for the kth highest probability. I would not characterize that as "incorrect" since this case is not really well-defined. And it's reasonable from an intuition pov since the tied tokens have equal likelihood.

So hopefully this PR can be simplified to remove that part?

It would be interesting to also test whether this is meaningfully faster on GPUs. I assume most of the overhead is the sort but if this is even slightly faster we might as well change the existing impl to do the count + mask approach.

hyeygit commented:
Re @njhill:

> The behaviour without doing this will already be effectively the same as an arbitrary tie-break.

Without random perturbation the returned p-set will include all the tie values which can be incorrect. Consider this toy example: given input probabilities `probs = [0.1, 0.2, 0.2, 0.2, 0.2]` and `p = 0.15`, the top-p cutoff probability will be `0.2`, and thresholding yields a mask of `probs < 0.2 == [False, False, False, False, False]` which means the returned top-p set is just `probs`. This is evidently incorrect since `sum(probs)==1` is much larger than `p==0.15`.

With the perturbation, a correct number of tie values are randomly chosen to be included in the p-set. This seems like a reasonable and intuitive behavior to me.

Regarding `apply_top_k_only`:
> Again this is no different to how the pre-existing impl works.

IIUC the pre-existing impl is exact and would return exactly `k` elements despite the presence of ties. 

> And I think it's ok for the actual shortlist to comprise more than k tokens in the case that there's a tie for the kth highest probability. I would not characterize that as "incorrect" since this case is not really well-defined.

Agree that it's intuitively acceptable to have a top-k set with more than `k` elements when there are ties. But from an API perspective it seems one could reasonably assume that a topk function shall return exactly `k` elements, especially since that's the behavior of `torch.topk` (documentation states that exactly `k` elements are returned).

---

Bottom line is, IMHO, that it _might_ be fine for top-k to not break ties, but not top-p. Ties must be correctly handled for top-p otherwise the returned p-set could be "too wrong".

njhill commented:
Thanks @hyeygit

> Consider this toy example:

Ah, I see what you mean now. I was thinking about the cumsum mask. However, following similar rationale to the top_k case, I actually still think this is fine/preferred. Specifically that we return a non-minimally sized set of tokens whose probabilities sum to > top_p in the case of ties. This is again because the tiebreaking is already nondeterministic and in theory any of the tied tokens could end up being chosen.

> IIUC the pre-existing impl is exact and would return exactly `k` elements despite the presence of ties.

I think you may have miunderstood the logic. It's masking based on the value of the gathered logits, not their indices: https://github.com/vllm-project/vllm/blob/e6e3c55ef28f30bca855399419c61bb70af03db2/vllm/v1/sample/ops/topk_topp_sampler.py#L167-L169

The scatter actually isn't needed at all for top_k. If your top_p impl is >= the speed of the existing impl on GPU too, I think we can remove scatter altogether.

>  especially since that's the behavior of `torch.topk` (documentation states that exactly `k` elements are returned).

While it's true that `torch.topk` will return exactly k, its tiebreaking is nondeterministic so in general any of the tied kth highest tokens could be chosen.

hyeygit commented:
Thank you @njhill .

> I think you may have miunderstood the logic. It's masking based on the value of the gathered logits, not their indices

You are right - I must've misread the original impl. Thanks for pointing it out!

> the tiebreaking is already nondeterministic and in theory any of the tied tokens could end up being chosen.

After thinking about it more, this now makes sense to me. In the case of ties, all tied values are included in the top-k / top-p set, and during the final sampling, all tie tokens have equal chance of being chosen. I guess we can consider the ties being broken during final sampling.

Let me remove the random perturbation. I'll just document the tie-non-breaking behavior in the docstring.

Thanks again for the review and discussion!

hyeygit committed:
Change tie-breaking behavior

* Do not break ties. Instead, include all tied tokens in the return set
  and leave tie breaking to the final sampling stage (since all tie tokens
  will have equal probability of being chosen).
* Removed random perturbation.
* Removed warning regarding the algorithm being approx.
* Edited tests.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

njhill commented:
Thanks @hyeygit!

>  In the case of ties, all tied values are included in the top-k / top-p set, and during the final sampling, all tie tokens have equal chance of being chosen. I guess we can consider the ties being broken during final sampling.

Yes, exactly. In addition such ties are relatively infrequent, and even less frequent that > 2 tokens would be tied for last place.

NickLucche commented:
Let's get the topk in so we can enable both :)

hyeygit commented:
> Let's get the topk in so we can enable both :)

Yep! I have the code ready, but since Github doesn't support chained PRs (when merging from a forked repo), I need to wait for this PR to be merged before being able to send out the Top-k one (or maybe there's a better way I am not aware of).

NickLucche commented:
I think you can just push to another branch and open a separate PR with topk so that I can rebase on top of that. Then you just wait for main to catch up. 
Anyways we should be able to get this stamped today if tests are passing.

hyeygit commented:
> I think you can just push to another branch and open a separate PR with topk so that I can rebase on top of that. Then you just wait for main to catch up. Anyways we should be able to get this stamped today if tests are passing.

@NickLucche Sounds good, I created a draft PR #15891

hyeygit commented:
[buildkite/ci/pr/v1-test](https://buildkite.com/vllm/ci/builds/16727#0195f1dd-9b2f-470e-a8fe-49a4e8820af2) is failing due to CUDA/XPU out of memory. Is there a way to re-run these tests?

yaochengji commented on tests/v1/tpu/test_topk_topp_sampler.py:
```diff
@@ -0,0 +1,120 @@
+
+        # Verify that the masked logit's probability sums to at least p.
+        probs.masked_fill_(logits_masked.isinf(), 0)
+        masked_prob_sum = probs.sum(dim=-1)
```
Usually it's a good idea to put a `xm.mark_step` before the assert to clear the graph

hyeygit replied:
Done.

yaochengji commented on tests/v1/tpu/test_topk_topp_sampler.py:
```diff
@@ -0,0 +1,120 @@
+        # Expect the smallest elements to be dropped.
+        expected_result = logits.clone()
+        expected_result[0, 0] = float("-inf")
+        expected_result[1, 1] = float("-inf")
```
I'd like to suggest put the expected result on cpu and compare two cpu tensors in the assert.

Also, we need to add a `xm.mark_stap` before converting the xla_tensor to cpu tensor.

hyeygit replied:
Done. Thanks for the review!

hyeygit committed:
Update tests to perform assertion on CPU.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

yaochengji commented on tests/v1/tpu/test_topk_topp_sampler.py:
```diff
@@ -0,0 +1,130 @@
+                                              k=no_op_k,
+                                              p=p)
+
+        xm.mark_step()
```
```
        xm.mark_step()
        # Verify that the masked logit's probability sums to at least p.
        probs.masked_fill_(logits_masked.isinf(), 0)
        masked_prob_sum = probs.sum(dim=-1)
        xm.mark_step()
        masked_prob_sum = masked_prob_sum.cpu()
        assert torch.all(torch.ge(masked_prob_sum + TOLERANCE, p))
```

Could you modify your code as above? It's not guaranteed to make the computation on cpu even though it is under a cpu context.

The basic way to make a clear boundary between TPU and CPU can be like this:
```
# computation on tpu
xm.mark_step()
cpu_a = tpu_a.cpu() # transfer necessary tpu tensor to cpu
cpu_b = tpu_b.cpu()
# computation on cpu
```

hyeygit replied:
Done. Added explicit `.cpu()` calls.

NickLucche commented:
Are there any more issues with tests on this PR?
`VLLM_XLA_CACHE_PATH= VLLM_XLA_CHECK_RECOMPILATION=1 VLLM_USE_V1=1 python -m pytest -s -v tests/v1/tpu/test_topk_topp_sampler.py` is green on my side here.

hyeygit committed:
Update tests to use explicit .cpu() calls.

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

yaochengji reviewed (approved):
LGTM, thanks!

mgoin commented:
Please see the failing v1 test, it looks like the cuda sampler tests are failing https://buildkite.com/vllm/ci/builds/16819/steps?jid=0195f7bd-4a4f-4517-af78-dc4a6772ba71

hyeygit committed:
Fix merge mistake

Signed-off-by: Hyesoo Yang <hyeygit@gmail.com>

njhill commented on vllm/v1/sample/ops/topk_topp_sampler.py:
```diff
@@ -199,7 +224,7 @@ def apply_top_k_only(
     # topk.values tensor has shape [batch_size, max_top_k].
     # Convert top k to 0-based index in range [0, max_top_k).
-    k_index = k.sub_(1).unsqueeze(1)
+    k_index = k.sub_(1).unsqueeze(1).expand(logits.shape[0], 1)
```
@hyeygit is this because of a TPU torch broadcasting limitation?

hyeygit replied:
Yes I think so. Without the explicit expand this fails on XLA.

Since TPU no longer uses this code path, I can revert this change if needed.

hyeygit commented:
> Please see the failing v1 test, it looks like the cuda sampler tests are failing https://buildkite.com/vllm/ci/builds/16819/steps?jid=0195f7bd-4a4f-4517-af78-dc4a6772ba71

Oh this is probably caused by my incorrect rebase -- had some duplicate lines in the sampler. After resolving the conflicts the tests seem to pass.

Thanks for the approval!
